File: src/main/java/com/bot/utils/CommandParsingUtils.java
package com.bot.utils;

import com.bot.models.Scroll;
import kotlin.Pair;

import java.util.ArrayList;
import java.util.List;

public class CommandParsingUtils {

    public static List<Pair<Scroll, Integer>> parseScrollUpdates(String input) throws IllegalArgumentException {
        var scrolls = input.split(",");

        // Scroll pair denotes a pair of scroll/number
        List<Pair<Scroll, Integer>> scrollPairs = new ArrayList<>();

        for (String scrollPair : scrolls) {
            var parts = scrollPair.split(" ");
            if (parts.length != 2) {
                throw new IllegalArgumentException("Error found near `" +
                        scrollPair + "` More than 2 parts found.");
            }
            // Find the scroll name and the quantity
            // TODO: Custom exceptions
            int quantity;
            Scroll scroll;
            if (StringUtils.isNumeric(parts[0])) {
                quantity = Integer.parseInt(parts[0]);
                scroll = Scroll.getScrollForName(parts[1]);
            } else if (StringUtils.isNumeric(parts[1])) {
                quantity = Integer.parseInt(parts[1]);
                scroll = Scroll.getScrollForName(parts[0]);
            } else {
                throw new IllegalArgumentException("Error found near `" +
                        scrollPair + "` No quantity found.");
            }
            if (scroll == null) {
                throw new IllegalArgumentException("Error found near `" +
                        scrollPair + "` No valid scroll name found.");
            }
            scrollPairs.add(new Pair<>(scroll, quantity));
        }
        return scrollPairs;
    }
}

File: src/main/java/com/bot/utils/CommandParsingUtils.kt
package com.bot.utils

import java.lang.IllegalArgumentException
import com.bot.models.Scroll
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

object CommandParsingUtils {
    @Throws(IllegalArgumentException::class)
    fun parseScrollUpdates(input: String): List<Pair<Scroll, Int>> {
        val scrolls = input.split(", ").toTypedArray()

        // Scroll pair denotes a pair of scroll/number
        val scrollPairs: MutableList<Pair<Scroll, Int>> = ArrayList()
        for (scrollPair in scrolls) {
            val parts = scrollPair.split(" ").toTypedArray()
            require(parts.size == 2) {
                "Error found near `" +
                        scrollPair + "` More than 2 parts found."
            }
            // Find the scroll name and the quantity
            // TODO: Custom exceptions
            var quantity: Int
            var scroll: Scroll?
            if (StringUtils.isNumeric(parts[0])) {
                quantity = parts[0].toInt()
                scroll = Scroll.getScrollForName(parts[1])
            } else if (StringUtils.isNumeric(parts[1])) {
                quantity = parts[1].toInt()
                scroll = Scroll.getScrollForName(parts[0])
            } else {
                throw IllegalArgumentException("Error found near `" +
                        scrollPair + "` No quantity found.")
            }
            requireNotNull(scroll) {
                "Error found near `" +
                        scrollPair + "` No valid scroll name found."
            }
            scrollPairs.add(Pair(scroll, quantity))
        }
        return scrollPairs
    }

    fun parseArgsToDate(input: String, warStartTime: String) : Date? {
        val args = input.split(" ")
        val formatter = SimpleDateFormat("dd-M-yyyy hh:mm:ss", Locale.ENGLISH)
        formatter.timeZone = TimeZone.getTimeZone("America/Chicago")
        var date: Date? = null
        for (arg in args) {
            val timeToParse = "$arg $warStartTime"
            try {
                date = formatter.parse(timeToParse)
                break
            } catch (e: ParseException) {
                continue
            }
        }
        return date
    }
}